// client/src/hooks/highJewelleryDummyData.js
export const  highJewelleryDummyData = [
  {
    _id: "hj001",
    name: "Royal Emerald Necklace",
    type: "high-jewellery",
    category: "necklaces",
    heroImage: "https://images.unsplash.com/photo-1515562141207-7a88fb7ce338?w=400",
    gallery: [
      "https://images.unsplash.com/photo-1515562141207-7a88fb7ce338?w=400",
      "https://images.unsplash.com/photo-1506630448388-4e683c67ddb0?w=400"
    ],
    shortDescription: "Exquisite emerald necklace with 18k gold setting featuring rare Colombian emeralds",
    colors: ["#FFD700", "#C0C0C0", "#E7B9A4"],
    slug: "royal-emerald-necklace",
    isHighJewellery: true
  },
  {
    _id: "hj002", 
    name: "Diamond Tiara Collection",
    type: "high-jewellery",
    category: "tiaras",
    heroImage: "https://images.unsplash.com/photo-1605100804763-247f67b3557e?w=400",
    gallery: [
      "https://images.unsplash.com/photo-1605100804763-247f67b3557e?w=400"
    ],
    shortDescription: "Handcrafted tiara with over 500 diamonds in platinum setting",
    colors: ["#E5E4E2", "#FFD700"],
    slug: "diamond-tiara-collection",
    isHighJewellery: true
  },
  {
    _id: "hj003",
    name: "Sapphire Statement Earrings",
    type: "high-jewellery", 
    category: "earrings",
    heroImage: "https://images.unsplash.com/photo-1535632066927-ab7c9ab60908?w=400",
    gallery: [
      "https://images.unsplash.com/photo-1535632066927-ab7c9ab60908?w=400"
    ],
    shortDescription: "Blue sapphire chandelier earrings with micro-pavé diamonds",
    colors: ["#0F52BA", "#FFD700", "#C0C0C0"],
    slug: "sapphire-statement-earrings",
    isHighJewellery: true
  },
  {
    _id: "hj004",
    name: "Ruby Eternal Ring",
    type: "high-jewellery",
    category: "rings", 
    heroImage: "https://images.unsplash.com/photo-1605100804763-247f67b3557e?w=400",
    gallery: [
      "https://images.unsplash.com/photo-1605100804763-247f67b3557e?w=400"
    ],
    shortDescription: "Pigeon blood ruby ring with halo diamond setting in rose gold",
    colors: ["#E91E63", "#FFD700", "#E7B9A4"],
    slug: "ruby-eternal-ring",
    isHighJewellery: true
  }
];
